### ISSUE TITLE
Possible double-return to ArrayPool in JsonDocument.Dispose

### ISSUE THREAD

**Issue by GrabYourPitchforks** [dotnet/corefx#39930]
Related to https://github.com/dotnet/corefx/issues/36152

If the method `JsonDocument.Dispose` is called from multiple threads concurrently, the risk exists that the underlying `_extraRentedBytes` and `_parsedData._data` arrays may be released back to `ArrayPool<byte>.Shared` multiple times, which could cause application-wide memory corruption. Guidance has generally been that improper multi-threaded access to instances which are not marked as thread-safe is allowed to corrupt the target instance itself, but it shouldn't corrupt global application state.

/cc @bartonjs, who has been looking at ArrayPool usage throughout the Framework.

**Comment by bartonjs** [comment]
Ugh, I thought I fixed it to Interlocked.Exchange with null.

**Comment by joshlang** [comment]
Ditto here:

https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/Common/src/System/Text/ValueUtf8Converter.cs
(in dispose)

https://github.com/dotnet/corefx/blob/94c5f1384c9b6477d164050de964cf6dcf99b2a9/src/System.Text.Json/src/System/Text/Json/Serialization/PooledByteBufferWriter.cs
(in dispose, but also CheckAndResizeBuffer)

https://github.com/dotnet/corefx/blob/cd020c11b563b51f439324bcefc559d1408f4e09/src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.MetadataDb.cs
(in TrimExcess, Dispose AND Enlarge)

https://github.com/dotnet/corefx/blob/15abdfc3816352c64098735b79a98dd97524ee36/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ArrayBuffer.cs
(in Dispose, and EnsureAvailableSpace)

https://github.com/dotnet/corefx/blob/bca44089343cf14cffd3cb0982b7e209692f6731/src/System.IO.Compression.Brotli/src/System/IO/Compression/BrotliStream.cs
(in ReleaseStateForDispose)

https://github.com/dotnet/corefx/blob/cece0df13a74d188f96b08ae5af0cd4cd60f8f76/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs
(in Dispose and DisposeAsyncCore and CopyFromSourceToDestinationAsync and CopyFromSourceToDestination)

https://github.com/dotnet/corefx/blob/2469a91c2aa68f9cb6b0f18f75dffc673a1f7895/src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs
(in Release)

https://github.com/dotnet/corefx/blob/1841042b99062de13dc80098cede9413be569238/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs
(in Dispose)

https://github.com/dotnet/corefx/blob/1841042b99062de13dc80098cede9413be569238/src/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
(CloseInternal might be okay but I wouldn't trust the _nestedRead flag so I'd still fix the general pattern anyway)

https://github.com/dotnet/corefx/blob/e7134de2b56bd1c4af0b5c58fe6e2cfdb8b72e5a/src/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs
(in ResetMemory... and those comments make me nervous)

https://github.com/dotnet/corefx/blob/1841042b99062de13dc80098cede9413be569238/src/System.Net.Http/src/System/Net/Http/HttpContent.cs
(in Dispose and in Grow)

https://github.com/dotnet/corefx/blob/af562e84b8b5a4a335ceb5006886a6c42c1c023e/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs
(in Clear)

https://github.com/dotnet/corefx/blob/105b70a2ee48cca52fef404449f350a4be485b40/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs
(in SendTransferState.Dispose via Cleanup)

https://github.com/dotnet/corefx/blob/2e73965a126946afc986175fd667ce3c31f03731/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs
(in CopyToContentLengthAsync ... technically probably okay, but why not fix the pattern anyway)

https://github.com/dotnet/corefx/blob/64609475281f2be05a0b9ca5b7bbdaf43b12dc55/src/Common/src/System/Security/Cryptography/Asn1Reader/AsnWriter.cs
(in Dispose and EnsureWriteCapacity via CryptoPool.Return)

Probably all should be Interlocked.Exchange -> null

**Comment by bartonjs** [comment]
Many of the additional ones listed are safe by usage: there's provably only one thread touching the object at a time. (JsonDocument.MetadataDb trim/enlarge, AsnWriter, the chain processor WorkingChain structure, etc)

**Comment by joshlang** [comment]
Forever?

**Comment by ahsonkhan** [comment]
> > Ugh, I thought I fixed it to Interlocked.Exchange with null.

@bartonjs, is that the suggested fix then? Use interlocked.exchange to make the dispose/return atomic?

**Comment by bartonjs** [comment]
```diff
            // When "extra rented bytes exist" they contain the document,
            // and thus need to be cleared before being returned.
-           if (_extraRentedBytes != null)
+           byte[] extraRentedBytes = Interlocked.Exchange(ref extraRentedBytes, null);
+
+           if (extraRentedBytes != null)
```

Similarly in MetadataDb.Dispose (but Grow isn't needed).

It's any Dispose on any type that's ever public or assigned to a field.  Growing things are differently dangerous under multi-threading; but double-return from parallel dispose is really, really bad.

AsnWriter has this pattern, but it's not public and never a field; only a local or method parameter.  I don't remember the rest off the top of my head.

**Comment by ahsonkhan** [comment]
Re-opening until its fixed in 3.0: https://github.com/dotnet/corefx/pull/40324

**Comment by ahsonkhan** [comment]
Fixed in 3.0 in https://github.com/dotnet/corefx/pull/40324.
